src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/AbstractRowsEventDataDeserializer.java

```diff
@@ -354,7 +353,14 @@ protected Serializable deserializeDatetimeV2(int meta, ByteArrayInputStream inpu
         if (deserializeDateAndTimeAsLong) {
             return castTimestamp(timestamp, fsp);
         }
-        return timestamp != null ? new java.util.Date(timestamp) : null;
+
+        return timestamp != null ? convertLongTimestamptWithFSP(timestamp, fsp) : null;
```

Comment:
We're returning a java.sql.Timestamp now instead of a java.util.Date. If you look at the implementation you'll see that java.sql.Timestamp is a light wrapper around java.util.Date, adding microsecond precision that java.util.Date lacks.